reconcile/dashdotdb_dora.py

```diff
@@ -0,0 +1,480 @@
+                )
+
+                result = cur.fetchall()
+
+                if not result:
+                    break
+
+                records.extend(result)
+
+                offset += limit
+
+            if not records:
+                LOG.info("No deployments found for %s", app_env)
+
+            return [Deployment(row[0], row[1]) for row in records]
+
+
+@dataclass
+class Commit:
+    repo: str
+    sha: str
+    date: datetime
+
+    def lttc(self, finish_timestamp: datetime) -> int:
+        commit_date_tzaware = self.date
+        finish_timestamp_tzaware = finish_timestamp
+
+        if commit_date_tzaware.tzinfo is None:
+            commit_date_tzaware = commit_date_tzaware.replace(tzinfo=timezone.utc)
+
+        if finish_timestamp_tzaware.tzinfo is None:
+            finish_timestamp_tzaware = finish_timestamp_tzaware.replace(
+                tzinfo=timezone.utc
+            )
+
+        return int((finish_timestamp_tzaware - commit_date_tzaware).total_seconds())
+
+
+@dataclass(eq=True, frozen=True)
+class SaasTarget:
+    app_name: str
+    env_name: str
+    path: str
+    resource_template: str
+    namespace: str
+    pipeline: str
+
+    @property
+    def app_env(self) -> AppEnv:
+        return AppEnv(self.app_name, self.env_name)
+
+
+@dataclass(eq=True, frozen=True)
+class RepoChanges:
+    repo_url: Optional[str]
+    ref_from: Optional[str]
+    ref_to: Optional[str]
+
+
+@dataclass(eq=True, frozen=True)
+class SummaryKey:
+    rt_name: str
+    target_ns: str
+
+
+@dataclass(eq=True, frozen=True)
+class SummaryEntry:
+    repo_url: str
+    target_ref: str
+
+
+class DashdotdbDORA(DashdotdbBase):
+    def __init__(self, dry_run: bool, thread_pool_size: int = 5) -> None:
+        settings = queries.get_app_interface_settings()
+        vault_settings = get_app_interface_vault_settings()
+        secret_reader = create_secret_reader(use_vault=vault_settings.vault)
+
+        # init dashdotdb
+        super().__init__(
+            dry_run=dry_run,
+            thread_pool_size=thread_pool_size,
+            marker="DDDB_DORA:",
+            scope="dora",
+            secret_reader=secret_reader,
+        )
+
+        # init dep db
+        depdb_secret = secret_reader.read_all_secret(DEPLOYMENT_DB_SECRET)
+        self.dep_db = DeploymentDB(
+            depdb_secret["db.host"],
+            depdb_secret["db.port"],
+            depdb_secret["db.name"],
+            depdb_secret["db.user"],
+            depdb_secret["db.password"],
+        )
+
+        # init github secret
+        gh_instance = queries.get_github_instance()
+        self.gh_token = secret_reader.read(gh_instance["token"])
+
+        # init app_interface_get_file
+        gl_instance = queries.get_gitlab_instance()
+        gl = GitLabApi(
+            gl_instance, project_id=13582, settings=settings
+        )  # 13582 is app-interface
+
+        self.app_interface_get_file = functools.cache(gl.get_file)
```

Comment:
I think this is not applicable because these DashdotDB reconcilers run on a cron expression.

Reply:
An easy fix is to use instance level cache, as the runtime may change from a cronjob to a long running deployment.

```python
self.app_interface_get_file = functools.cache()(gl.get_file)
```

Reply:
I don't understand `functools.cache()(gl.get_file)`, it also seems to fail:

```
>>> import functools
>>> functools.cache()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: cache() missing 1 required positional argument: 'user_function'
>>> def a(s):
...     return s+"!"
...
>>> functools.cache()(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: cache() missing 1 required positional argument: 'user_function'
```

Reply:
oh, sorry, the signature is a little different than `lru_cache`, it' OK, your usage is already good!

source: https://github.com/python/cpython/blob/afa7b0d743d9b7172d58d1e4c28da8324b9be2eb/Lib/functools.py#L652

```python
def cache(user_function, /):
    'Simple lightweight unbounded cache.  Sometimes called "memoize".'
    return lru_cache(maxsize=None)(user_function)
```

So it's already a instance level cache! Use `@cache` on instance method will cause memory leak because the scope is different, use normal function call is good here.